src/util/gc_byte.rs

```diff
@@ -1,11 +1,35 @@
-use crate::util::ObjectReference;
+use crate::util::side_metadata::{SideMetadata, SideMetadataID};
+use crate::util::{constants, ObjectReference};
 use crate::vm::ObjectModel;
 use crate::vm::VMBinding;
 use std::sync::atomic::{AtomicU8, Ordering};
 
+use super::Address;
+
 /// This struct encapsulates operations on the per-object GC byte (metadata)
 pub struct GCByte {}
 
+// For performance reasons, mutable static is used over lazy_static or sync::once.
+// This requires several unsafe blocks in the current file, but it looks safe to use these unsafe blocks, because we only assign the mutable static once.
+//
+// NOTE: A more suitable implementation may also be possible.
+static mut SIDE_GCBYTE_ID: SideMetadataID = SideMetadataID::new();
+
+#[allow(clippy::cast_ref_to_mut)]
+#[allow(clippy::mut_from_ref)]
+pub(crate) unsafe fn init_side_gcbyte() {
+    let res = SideMetadata::request_meta_bits(
+        // constants::BITS_IN_BYTE,
+        2,
```

Comment:
2-bits is the number of bits we are currently using inside the GC byte.
The other option I had for specifying the number of bits was 8, which leaves 6-bits unused.

Reply:
Yeah, I understand that we use 2 bits in the GC byte (for SemiSpace). Is it defined here?
https://github.com/mmtk/mmtk-core/blob/bb9fd6c9acfd685af729ca6be9a9d3cb9b26435f/src/plan/semispace/constraints.rs#L5

If so, that means it depends on the plan. We cannot assume we always use 2 bits.

Reply:
If I remember correctly, this constant has never been used in the GC byte's implementation.
In other words, the current hard-coded 2-bits GC byte implementation is equivalent to the in-header GC byte implementation.
If we want to generalise GC bytes implementation, I think it's better to do it in a separate PR.

Reply:
> In other words, the current hard-coded 2-bits GC byte implementation is equivalent to the in-header GC byte implementation.

Can you point me to the place where we assume 2 bits in the in-header GC byte? 

I still think this is a good place to use the constant. And the bits is plan-specific. E.g. for NoGC, because the constant is 0, no metadata side table is needed.

Reply:
You're right, we don't make that assumption anywhere, but we only use 2-bits for the forwarding word in the currently available plans.
For NOGC, we don't initialise the GC bytes (e.g. it's only called in CommonPlan).

I think it's easy to add the constant here, I'm just not sure how it will work for the new PR where runtime plan selection is possible.

Reply:
If the 2bits is not plan specific, then refactoring for runtime plan selection won't touch it. If it is plan specific, then you should use `SelectedConstraints`, and it will be turned to https://github.com/mmtk/mmtk-core/blob/0fc3a168a0f97bac9659e6fe856a5bdf8b3cc294/src/plan/plan_constraints.rs#L7.

Reply:
I assume that future plans may need more than 2 bits and the number of bits is plan-specific.
My concern is that if the selected plan changes at runtime, the number of bits may also change.
This will require reinitialisation of the side GC byte, mainly because it changes the amount of memory required for the side GC byte.
E.g. if the required memory increases(say from 2bits to 4bits), the whole space may not fit in the previous address range, if a policy-specific side metadata is located after the side GC byte (we organise the side metadata spaces sequentially).

One simple (but in-efficient) solution would be to always consider the side GC byte to be 8-bits.
To make this more efficient, the side metadata module will need a new interface which specifically allows changing the number of bits in the side GC byte at runtime.
However, I'm not sure this is necessary/useful.

Reply:
By runtime plan selection, I mean allowing selecting a plan at boot time. They cant switch plans once a plan is selected. So we won't reinitialise side GC byte due to plan. There will be only one-time initialisation for side GC byte.

Reply:
We don't need to overthink on this. My point is simple:
1. Use a constant instead of a magic number `2`
2. If the constant is plan specific, we should use `SelectedConstraints`, as that is the intended way for plan specific constants.

Reply:
If the 2bits is not plan specific, then just use a constant for the 2bit. For plans that need extra bits, they can do their own calculation (e.g. `GC_HEADER_BITS - GLOBAL_GC_BITS`, and create their own side metadata for the extra bits).